src/fm/utils.cljc

```diff
@@ -249,18 +267,23 @@
            (s/valid? :fm/anomaly ~ret-sym)
            (~handler-sym ~ret-sym)
 
-           ~@(if (:fm/ret metadata)
-               [`(s/valid? ~ret-spec-sym ~ret-sym)
-                ret-sym
-
-                :else
+           ~@(when (:fm/ret metadata)
+               [`(not (s/valid? ~ret-spec-sym ~ret-sym))
                 `(~handler-sym
                   #:fm.anomaly{:spec :fm.anomaly/ret
                                :sym  '~sym
                                :args ~args-syms
-                               :data (s/explain-data ~ret-spec-sym ~ret-sym)})]
+                               :data (s/explain-data ~ret-spec-sym ~ret-sym)})])
+
+           ~@(when (:fm/rel metadata)
```

Comment:
Does this mean if both ret and rel are provided both will be checked even if ret fails? I’m guessing that’s a good thing to inform callers of the multiple potential issues in their fn, but if the handler is invoked twice it seems like that could have unintended consequences (pertaining to the return value of the handler). Or am I overthinking this?

Reply:
The `:fm/ret` and `:fm/rel` paths can be treated totally independently, which took me a bit of experimentation to discover. So both are independently optional (`:fm/rel` is a actually superset of `:fm/ret`, but I think there's value in supporting both). In any case, the handler will only be invoked at most once, and its argument will depend on what branch of `cond` is taken.

This commit inverts the `cond` predicates, so the `:else` case represents "success", and everything else is a failure mode (i.e. invokes the handler) :joy:. This makes the branches a little more uniform, and made it simpler to express "optional" branches (i.e. branches that depend on the presence of `:fm/ret` or `:fm/rel`), but it also depends heavily on the predicate ordering.

If `:fm/ret` is provided, it is always checked first (L271). If the `:fm/ret` check fails, the handler receives `:fm.anomaly/ret` just as before, so that path is unchanged. If it succeeds, and `:fm/rel` is provided, it will go on to enforce `:fm/rel`. If `:fm/rel` isn't provided, it will reach `:else ~ret-sym` i.e. "succeed".

If `:fm/rel` is provided and `:fm/ret` either passes or isn't provided, L279 will enforce `:fm/rel`. If `:fm/rel` doesn't hold, the handler is invoked with `:fm.anomaly/rel`, which is a new path (with a new anomaly, etc., etc.). If `:fm/rel` holds, it will reach `:else ~ret-sym` and start dancing and cheering.

This would be much cleaner as a decision tree, but I'm pretty garbage at making those. Hopefully this wall of text is at least somewhat helpful.

Reply:
This is another way of interpreting those `when` forms:
```
~@(when meta-present?
    [cond-predicate-form
     cond-result-form])
```
It's conditionally emitting a `cond` branch, which is a little confusing no matter how you slice it.

Reply:
Ah, that totally makes sense. I should’ve opened it on my computer, I couldn’t tell it was part of a cond. Perfect solution for me.